### PR TITLE
SOL-153443: report broker admission pressure in the log

### PR DESCRIPTION
**Pull request by solace-aross**
Stacked on **`aross/SOL-153442-ratelimiter-fail-fast`** (PR #360), not `main`. That branch has not merged yet, so its commit shows up in this diff. Review only `cb5fc65`; the file-level diff against the base branch is this change alone.

## What and why

Support and SRE had no way to tell "the server is pacing or shedding to protect this broker" from "something else is slow". The only evidence was a per-request `slog.Debug` on the admitted path, plus SOL-153442's shed warning, which fires only once a request has already failed.

Two signals, both gated on `OBS_SATURATION_EVENTS_ENABLED` (default off), which until now gated nothing:

| Signal | When | Fields |
|---|---|---|
| `broker admission slow: request still waiting to be admitted` (WARN) | Once per request whose admission wait passes `observability.saturation_threshold_ms` | `broker`, `operation`, `stage`, `waited`, `threshold`, `max_queue_wait` |
| `broker in-flight occupancy` (INFO / WARN at the cap) | Every `observability.otel_self_stats_interval_s`, per broker carrying load | `broker` (alias), `in_flight`, `limit` |

Both reuse SOL-153442's `stage` vocabulary (`rate_limit`, `concurrency`) so they and the shed warning never disagree about which gate is involved.

## This is an interim signal, not the metric

The ticket originally asked for `mcp_saturation_total` and an occupancy gauge on the existing OpenTelemetry `/metrics` endpoint. That endpoint does not exist. There is no `/metrics` route, no meter provider, no exporter, and no OpenTelemetry or Prometheus dependency in `go.mod`; `internal/observability/metrics` is two boolean accessors, and `docs/observability.md:9` says so outright. Building the substrate here would duplicate work already in flight under SOL-150254.

So this ships as logs and cross-links. `docs/observability.md` now marks the capability **[Interim — logs only]** rather than shipped, and says plainly that the metric form is still roadmap. Neither log line is a stable interface; the compatibility commitment will live on the metric that replaces them.

## Two decisions worth a look

**The slow warning fires during the wait, not on resolution.** A resolution-time line arrives up to `max_queue_wait` later, and under `max_queue_wait: 0` (supported, restores unbounded waiting) it never arrives at all, which is the case that most needs a signal. Firing from inside the gate's `select` means the arm has to be re-entered, hence the labeled loops in `admit`; setting the timer channel to `nil` after it fires keeps this to one line per request rather than a flood that gets the signal switched off. SOL-153442's shared admission budget is untouched. The new timer is a second one-shot channel alongside `expired`, measured from the same entry point.

**`observability.saturation_threshold_ms` moves from `10` to `1000`** and is redefined from tool-call latency to admission wait. Nothing consumed it before, so no behavior changes. Against the default `100ms` `request_min_interval` the old value would have reported effectively every request as saturated.

## Testing

`make check` passes: build, vet, golangci-lint, race tests, 89.6% coverage against the 85% floor. New code is at 100% statement coverage apart from a three-line extracted helper.

Beyond the assertions, three mutation checks confirm the tests couple to the behavior rather than passing regardless:

- re-arming the slow timer instead of nil-ing it → `WarnsOnlyOnce` fails (6 warnings, want 1)
- swapping `len(sem)`/`cap(sem)` → the occupancy tests fail
- making the pool ignore the flag → `SaturationEventsEnabled_ReachesTheSender` fails

The "fires during the wait" test asserts the warning has appeared *while* the sender is still parked in `admit`, so a resolution-time implementation cannot pass it. Timing tests were run at `-count=2 -cpu=1` under `-race` to check behavior on a constrained runner.

## Not in scope

Dashboards and alerting rules, and the metric form itself (SOL-150254).

### CODE REVIEW

**Comment by solace-aross** [comment]
Pushed `44113dd` addressing a skeptical review pass. One HIGH finding was real and mutation-proven.

**HIGH — the concurrency-gate test asserted the log line but not the gate.** Adding `break concurrencyGate` to the warn arm made every slow request bypass `semp.max_concurrent_per_broker` entirely, plus leak a slot the deferred release never accounted for, and the full suite still passed. Unbounded concurrency against exactly the degraded broker this feature protects. Fixed with the missing non-blocking check on that test, plus a warn-then-succeed test that frees the slot and asserts `admit` returns nil holding exactly one slot. Both fail against the mutant now, verified.

That also closed a coverage gap worth naming: every warning test previously either parked forever or shed, so the "fires regardless of outcome" requirement only had its shed half tested.

**MEDIUM**
- The 1000ms default is really sized against `fanOutDefaultConcurrency` (8) × `request_min_interval` (100ms) ≈ 800ms, not the pacing interval alone. A 20% margin that nothing recorded. Now named in the `defaults.go` reasoning and in `docs/observability.md`, with the consequence spelled out: raise fan-out concurrency and healthy fan-out steps become a WARN flood.
- The periodic line is a timer sample that skips zero, so it structurally misses bursts shorter than the interval while the doc claimed it answers "which broker is under pressure right now". Now documented as detecting sustained pressure, with the per-request warning as the transient catcher.

**LOW**
- Occupancy now reports once at startup instead of waiting a full 60s interval, so flipping the flag on mid-incident is useful immediately.
- The `interval <= 0` guard is now pinned by a test; it stands between a caller mistake and a `time.NewTicker` panic on a background goroutine.
- Corrected the stale "sole reader/writer of `p.clients`" comment on `getOrCreate`.

**Verified clean by the reviewer, no action:** SOL-153442's shared admission budget survives the labeled-loop rework intact (no spin, no double-acquire, no missed `expired` arm, no changed error path, and the Go 1.23 timer-channel guarantee holds identically for the new timer); reporter lifecycle and the LIFO defer ordering in `main.go`; `len(sem)`/`cap(sem)` genuinely race-free rather than a race the detector missed; no lock hazard in `OccupancySnapshot`; secure logging; and the threshold change consuming nothing prior.

Flakiness was checked independently at `-race -count=60 -cpu=1` under 8 background CPU hogs, and held.

`make check` still passes at 89.6% coverage.

**Comment by solace-aross** [comment]
**CI caveat on this stacked PR, worth knowing before you read the green checkmarks.**

`build-and-test.yml` scopes its `pull_request` trigger to `branches: [main]`, deliberately, so a PR against a throwaway base does not spin up five Dockerized E2E suites. This PR targets `aross/SOL-153442-ratelimiter-fail-fast`, so the build, lint, unit-test and E2E jobs did **not** run from the PR event. The 15 green checks here are the auxiliary ones: DCO, CHANGELOG, licences, SBOM, Jira, and the script self-tests.

I have dispatched the full suite against the branch by the route the workflow's own comment documents for exactly this case:

    gh workflow run build-and-test.yml --ref aross/SOL-153443-saturation-metric

Run: https://github.com/SolaceProducts/solace-broker-mcp/actions/runs/33557929711 — I will post the verdict here.

Two follow-ups for whoever merges:
- Retarget this PR to `main` once #360 lands. That is what makes the required contexts report on the PR itself.
- Do not merge on the strength of the 15 checks above alone; they do not include a compiler.

Locally, `make check` passes: build, vet, golangci-lint, race-enabled tests, 89.6% coverage against the 85% floor.

**Comment by solace-aross** [comment]
Full suite is green on the branch. Run: https://github.com/SolaceProducts/solace-broker-mcp/actions/runs/33557929711

| Job | Result |
|---|---|
| build | pass |
| lint | pass |
| e2e-basic-mcp | pass |
| e2e-oauth | pass |
| e2e-monitoring | pass |
| e2e-management | pass |
| e2e-action | pass |

Plus the 15 auxiliary checks on the PR itself. Ready for review.

**Review by bczoma — DISMISSED** [review]
## Review summary

Wires `OBS_SATURATION_EVENTS_ENABLED` to two new log signals (per-request slow-admission WARN, periodic per-broker occupancy log), reusing #360's `stage` vocabulary. The ticket literally asked for a real OTel `/metrics` counter+gauge; no metrics substrate exists in this codebase yet, so this ships as an explicitly-disclosed interim log signal instead — the pivot is stated consistently and accurately in the PR body, CHANGELOG, and `docs/observability.md` (four places), never overclaiming "shipped" for the metric.

I traced the `admit()` rework line-by-line against #360's pre-existing logic and confirmed the shared admission budget is genuinely untouched — no double-acquire, no spin, no missed timer arm. Your own self-review already caught and fixed a real HIGH finding (the concurrency-gate test gap) before I even started; verified that fix and its new test do catch the described mutant. All three mutation-testing claims in the PR body checked out against the actual tests.

### 🟡 Worth a look, non-blocking

- **CI coverage on this stacked PR**: since it targets a non-`main` branch, the real `build-and-test.yml` `pull_request` trigger never ran natively — only your manually-dispatched run is green. That dispatch isn't enforced by branch protection, so a later push (e.g. a rebase once #360 merges) wouldn't auto-retrigger it. Suggest treating "retarget to `main`, re-verify CI" as an explicit checklist item for whoever merges, not something to trust from the point-in-time manual run alone.
- **`internal/semp/resilience/sender.go` (`admit`)** — the slow-admission warning and the budget-expiry timer share one `select`; under a race, a request can shed without the warning ever firing (the shed WARN from #360 still fires, so visibility isn't fully lost — just this specific signal). Worth a one-line comment documenting this as a known trade-off.
- **`internal/observability/health/saturation.go:56-57`** — the in-code comment still claims the periodic line answers "which broker is under pressure right now," but `docs/observability.md` was already corrected (in your own self-review) to say it only detects sustained pressure, not bursts. Worth syncing the comment to match.

### 🟢 One thing I checked and downgraded

I initially flagged the Jira/AC mismatch (SOL-153443's literal `/metrics` ACs aren't met by this log-only interim signal, ticket sits in "Verifying") as a blocker, then checked it more carefully: "Verifying" is a reversible, non-terminal state with no auto-close on merge, and SOL-150254 already scopes the real `/metrics` work at the epic level. This is a ticket-hygiene item for whoever owns SOL-153443 (re-scope the ACs or link SOL-150254 before closing), fully decoupled from this PR's code — not a merge gate.

### Strengths

- Clean dependency inversion: `internal/observability/health` depends only on a function-shaped abstraction, never reaching into `semp`/`resilience` directly.
- Occupancy signal reads `len`/`cap` off the existing shared semaphore rather than adding a second counter that could drift — avoids a whole class of consistency bugs.
- Log field provenance for `broker` deliberately reuses #360's already-verified-safe pattern at the new call site.

**Recommendation: ✅ Approve** — nothing here blocks merge.

**Review by bczoma — DISMISSED** [review]
Re-approving after the retarget to `main` (now that #360 merged) and the rebase. Diffed the previously-reviewed commit against the new head — the only change is the `go.mod`/`go.sum`/`THIRD_PARTY_LICENSES.md` dependency bump picked up from rebasing (matches the merged kin-openapi Dependabot PR #365). No code changes to the reviewed feature.

File: CHANGELOG.md [hunk 1]
```diff
@@ -21,6 +21,8 @@ and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0
...
... whatever the outcome — and a request that goes on to be shed is still reported separately by the existing shed warning. `broker in-flight occupancy` fires every `observability.otel_self_stats_interval_s` (default `60s`) for each broker carrying load, carrying the broker alias, `in_flight` and the configured `limit`, at `WARN` when the two are equal and `INFO` below; idle brokers are skipped so the presence of a line is itself the signal, and brokers the server has never connected to do not appear, since broker clients are created on first use. **This is an interim signal, not the metric the observability roadmap described.** `docs/observability.md` listed `mcp_saturation_total` and a rate-limiter health gauge on a Prometheus `/metrics` endpoint; this build has no such endpoint, no meter provider and no OpenTelemetry dependency, and standing that pipeline up here would duplicate work already in flight under SOL-150254, so the metric form stays on the roadmap and that document now marks this capability `[Interim — logs only]` rather than shipped. Neither log line is a stable interface; the compatibility commitment will live on the metric that replaces them. Tracked under SOL-153443.
```
[the remark this answers is not in the document]
Comment (solace-aross) [reply CHANGELOG.md]: Fixed in e5383a4: reworded to say the /metrics endpoint exists (SOL-152091) but has no meter provider or saturation instruments registered on it yet, rather than claiming no endpoint exists.

File: docs/observability.md [hunk 2]
```diff
@@ -847,13 +855,105 @@ need to spend review time on them:
...
+> [Planned for a Later Release](#planned-for-a-later-release-not-frozen-in-this-review)
+> are still roadmap. Nothing in this section appears on `/metrics`, because this build has
+> no `/metrics` endpoint._
+
+Support and SRE needed one question answered quickly: when a customer says "MCP feels
+slow", is the server pacing or shedding requests to protect a broker, or is something else
+slow? The metric form of that answer depends on a metrics pipeline this build does not
+have — no endpoint, no meter provider, no OpenTelemetry dependency — and standing one up
+here would duplicate work already in flight under SOL-150254. So the signal ships as logs
+now and moves to metrics when that pipeline lands.
```
[the remark this answers is not in the document]
Comment (solace-aross) [reply docs/observability.md]: Fixed in e5383a4: reworded to say the /metrics endpoint exists (SOL-152091) but has no meter provider or saturation instruments registered on it yet, rather than claiming no endpoint exists.

File: internal/observability/health/saturation.go [hunk 3]
```diff
@@ -0,0 +1,132 @@
...
+
+// StartOccupancyReporter periodically logs how full each broker's in-flight
+// semaphore is, and returns a stop function that halts it.
+//
+// This is the interim, log-based half of SOL-153443. The observability schema
+// specifies this as a gauge on a Prometheus /metrics endpoint, but this build
+// has no such endpoint, no meter provider and no OpenTelemetry dependency, and
+// standing one up here would duplicate the work in flight under SOL-150254. A
+// periodic log line needs none of that and answers the same question: which
+// broker is under pressure right now. See docs/observability.md.
```
[the remark this answers is not in the document]
Comment (solace-aross) [reply internal/observability/health/saturation.go:58]: Fixed in e5383a4: reworded to say the /metrics endpoint exists (SOL-152091) but has no meter provider or saturation instruments registered on it yet, rather than claiming no endpoint exists.

File: internal/semp/resilience/sender.go [hunk 4]
```diff
@@ -136,6 +136,37 @@ type Sender struct {
...
+// single WARN per request that is still waiting to be admitted after slowAfter,
+// naming the gate it is stuck at.
+//
+// This is the stopgap for SOL-153443, not the metric the observability schema
+// describes. `docs/observability.md` specifies a `/metrics` endpoint that this
+// build does not have — there is no meter provider, no exporter, and no
+// OpenTelemetry dependency — and building one here would duplicate the
+// in-flight work under SOL-150254. A log line needs none of that and answers
+// the same first question: is the server pacing or shedding to protect this
+// broker, or is something else slow?
```
[the remark this answers is not in the document]
Comment (solace-aross) [reply internal/semp/resilience/sender.go:159]: Fixed in e5383a4: reworded to say the /metrics endpoint exists (SOL-152091) but has no meter provider or saturation instruments registered on it yet, rather than claiming no endpoint exists.

**Review by bczoma — APPROVED** [review]
Re-approving after the latest commit (e0695b4f). Reviewed it closely since it touches the exact "no /metrics endpoint" premise my earlier review relied on.

Verified: SOL-152091 (merged separately, #357) added a real `/metrics` endpoint (`startMetricsEndpoint` in `cmd/server/main.go`, confirmed present) — this PR's earlier text overclaimed "no such endpoint at all," which became stale/wrong once rebased onto main and briefly self-contradicted the CHANGELOG entry directly above it for #357. This commit corrects that in exactly four places (CHANGELOG.md, docs/observability.md, saturation.go's doc comment, sender.go's doc comment) to the accurate, narrower claim: the endpoint exists, but nothing registers a meter provider or saturation instruments on it yet. Confirmed doc/comment-only — no behavior change, matches the commit message.

This doesn't change any of the substantive findings from the earlier review (all code-level, unaffected by this wording fix) — good catch on your part, and CI is green (37/37) now that this properly targets `main`.
